librad/src/git/include.rs

```diff
@@ -0,0 +1,239 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    convert::TryFrom,
+    marker::PhantomData,
+    path::{self, PathBuf},
+};
+
+use git2::{Config, ConfigLevel};
+
+use crate::{
+    git::{
+        local::url::LocalUrl,
+        types::{remote::Remote, AsRefspec, FlatRef, Force, Reference},
+    },
+    hash::Hash,
+    meta::user::User,
+    peer::PeerId,
+    uri::RadUrn,
+};
+
+/// An `Include` is a representation of an include file which we want to
+/// generate for working copies.
+///
+/// When generated it will include a list of remotes of the form:
+/// ```text
+/// [remote "<handle>@<peer_id>"]
+///     url = rad://<repo>@<peer_id>.git
+///     fetch = refs/heads/*:refs/remotes/<repo>/*
+/// ```
+///
+/// This file can then be added to the working copy's `config` file as:
+/// ```text
+/// [include]
+///     path = <path>/<repo>.inc
+/// ```
+pub struct Include<Path> {
+    /// The list of remotes that will be generated for this include file.
+    pub remotes: Vec<Remote<LocalUrl>>,
+    /// The directory path where the include file will be stored.
+    pub path: Path,
+    /// The namespace this include file is interested in. In other words, the
+    /// generated include file should be for some project that lives under
+    /// this namespace in the monorepo.
+    ///
+    /// Note that the final file name will be named after
+    /// this namespace.
+    pub repo: Hash,
+}
+
+impl<Path> Include<Path> {
+    /// Create a new `Include` with an empty set of remotes.
+    pub fn new(path: Path, repo: Hash) -> Self {
+        Include {
+            remotes: vec![],
+            path,
+            repo,
+        }
+    }
+
+    /// Create the include file by creating a [`Config`].
+    pub fn create(self) -> Result<Config, git2::Error>
+    where
+        Path: AsRef<path::Path>,
+    {
+        git2::Config::try_from(self)
+    }
+
+    /// Convenience for adding an iterator of `Remote`s.
+    pub fn add_remotes(&mut self, remotes: impl Iterator<Item = Remote<LocalUrl>>) {
```

Comment:
If you put an `IntoIterator` bound here, the caller can pass a collection type or an iterator.

Reply:
oooh thanks!